test/e2e/operator.go

```
@@ -272,3 +277,102 @@ var _ = Describe("ARO Operator - Conditions", func() {
+		Expect(err).NotTo(HaveOccurred())
+		subnetMap := make(map[subnetDescriptor]bool)
+		for _, machine := range machines.Items {
+			_, subnetDesc, err := getDescriptorFromProviderSpec(machine.Spec.ProviderSpec.Value)
+			if err != nil {
+				log.Fatal(err)
+			}
+			subnetMap[*subnetDesc] = true
+		}
+		for ws := range subnetMap {
+			location = ws.location
+			resourceGroup = ws.resourceGroup
+			subnet = ws.subnetName
+			vnet = ws.vnetName
+		}
+	}
+	createDummyNSG := func() {
+		dummynsg = mgmtnetwork.SecurityGroup{
+			Location:                      to.StringPtr(location),
+			Name:                          to.StringPtr("dummynsg"),
+			Type:                          to.StringPtr("Microsoft.Network/networkSecurityGroups"),
+			SecurityGroupPropertiesFormat: &mgmtnetwork.SecurityGroupPropertiesFormat{},
+		}
+		err := clients.NetworkSecurityGroups.CreateOrUpdateAndWait(ctx, resourceGroup, "dummynsg", dummynsg)
+		Expect(err).NotTo(HaveOccurred())
+		dummynsgObject, err := clients.NetworkSecurityGroups.Get(ctx, resourceGroup, "dummynsg", "")
+		Expect(err).NotTo(HaveOccurred())
+		dummynsg.ID = dummynsgObject.ID
+	}
+	BeforeEach(func() {
+		enableReconcileNSG()
+		gatherNetworkInfo()
+		createDummyNSG()
+	})
+	AfterEach(func() {
+		err := clients.NetworkSecurityGroups.DeleteAndWait(context.Background(), resourceGroup, "dummynsg")
+		if err != nil {
+			log.Warn(err)
+		}
+	})
+	Specify("If worker nsg is changed or removed then controller should reconcile", func() {
+		validateNSG := func() (bool, error) {
+			s, err := clients.Subnet.Get(ctx, resourceGroup, vnet, subnet, "")
+			if err != nil {
+				return false, err
+			}
+			if *s.NetworkSecurityGroup.ID == *correctNSG {
+				log.Info("nsg matched expected value")
+				return true, nil
+			}
+			log.Info("nsg did not match expected value")
```

Nit: If the flow reach this end, what value would you think logging with `error` level and `*s.NetworkSecurityGroup.ID` information will bring? How much will it simplify debug with such information at hand?

---

I went with `info` at first because it is expected reconciler may need grace period before reconciling, but upon reflection of your comment I agree that any change of the network security group from expected group should throw an error. 

The `*s.NetWorkSecurityGroup.ID` and `*correctNSG` would definitely provide visibility into "why" it did not match expected value. I will update that.

---

I have added more descriptive and error level logging for when nsg are not equal